west/west.py

```diff
@@ -0,0 +1,170 @@
+"""
+This is where the migration plan is computed, by merging information
+from the existing files (west.files) and from the db (west.db)
+"""
+
+import io
+
+from west import db
+from west import files
+from west import utils
+from west.settings import settings
+
+
+def get_applied_versions():
+    """
+    Return the list of applied versions.
+    Reuse django migration table.
+    """
+    applied_versions = set(db.get_applied_versions())
+
+    known_versions = set(files.get_known_versions())
+
+    return utils.sortversions(applied_versions & known_versions)
+
+
+def get_closest_version(target_version, sql_tpl, existing_files,
+                        force_version=None):
+    """
+    Get the version of a file (schema or fixtures) to use to init a DB.
+    Take the closest to the target_version. Can be the same version, or older.
+    """
+    # get known versions
+    known_versions = files.get_known_versions()
+    # find target version
+
+    try:
+        previous_versions = list(utils.until(known_versions, target_version))
+    except ValueError:
+        raise ValueError(
+            'settings.TARGET_VERSION is improperly configured: '
+            'version {} not found.'.format(
+                settings.TARGET_VERSION))
+
+    # should we set a version from settings ?
+    if force_version:
+        if force_version not in previous_versions:
+            raise ValueError(
+                'settings.TARGET_VERSION is improperly configured: '
+                'settings.SCHEMA_VERSION is more recent.')
+
+        file = sql_tpl.format(force_version)
+        if file in existing_files:
+            return force_version
+
+        # not found
+        return None
+
+    for version in previous_versions[::-1]:
+        schema_file = sql_tpl.format(version)
+        if schema_file in existing_files:
+            return version
+
+    # not found
+    return None
+
+
+def get_schema_version():
+    """
+    Get the version to use to init a new DB to the current target version.
+    """
+    version = get_closest_version(
+        target_version=settings.TARGET_VERSION,
+        sql_tpl=settings.SCHEMA_TPL,
+        existing_files=files.get_known_schemas(),
+        force_version=settings.SCHEMA_VERSION)
+
+    if version is None:
+        raise ValueError('Can not find a schema to init the DB.')
+    return version
+
+
+def get_fixtures_version(target_version):
+    """
+    Get the closest fixtures to use to init a new DB
+    to the current target version.
+    """
+    version = get_closest_version(
+        target_version=target_version,
+        existing_files=files.get_known_fixtures(),
+        sql_tpl=settings.FIXTURES_TPL)
+
+    if version is None:
+        raise ValueError('Can not find fixtures to init the DB.')
+    return version
+
+
+def is_manual_migration(file_handler):
```

Comment:
Not exactly the same, but this function is defined also in the `files.py` module.
Is it possible to factorize ?

Reply:
Oh, I probably forgot do delete one of those